qualtran/bloqs/block_encoding/tensor_product.py

```diff
@@ -0,0 +1,154 @@
+    Register,
+    Signature,
+    SoquetT,
+)
+from qualtran.bloqs.block_encoding import BlockEncoding
+from qualtran.bloqs.block_encoding.lcu_select_and_prepare import PrepareOracle
+from qualtran.bloqs.bookkeeping import Partition
+from qualtran.symbolics import SymbolicFloat
+
+
+@frozen
+class TensorProduct(BlockEncoding):
+    r"""Tensor product of a sequence of block encodings.
+
+    Builds the block encoding as
+    $$
+        B[U_1 ⊗ U_2 ⊗ \cdots ⊗ U_n] = B[U_1] ⊗ B[U_2] ⊗ \cdots ⊗ B[U_n]
+    $$
+
+    When each $B[U_i]$ is a $(\alpha_i, a_i, \epsilon_i)$-block encoding of $U_i$, we have that
+    $B[U_1 ⊗ \cdots ⊗ U_n]$ is a $(\prod_i \alpha_i, \sum_i a_i, \sum_i \alpha_i \epsilon_i)$-block
+    encoding of $U_1 ⊗ \cdots ⊗ U_n$.
+
+    Args:
+        U: A sequence of block encodings.
+
+    Registers:
+        system: The system register.
+        ancilla: The ancilla register.
+        resource: The resource register.
+
+    References:
+        [Quantum algorithms: A survey of applications and end-to-end complexities](https://arxiv.org/abs/2310.03011). Dalzell et al. (2023). Ch. 10.2.
+    """
+
+    U: Sequence[BlockEncoding] = field(
+        converter=lambda x: x if isinstance(x, tuple) else tuple(x), validator=validators.min_len(1)
+    )
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature.build_from_dtypes(
+            system=self.dtype, ancilla=QAny(self.num_ancillas), resource=QAny(self.num_resource)
+        )
+
+    @cached_property
+    def dtype(self) -> QDType:
+        return QAny(bitsize=sum(u.dtype.num_qubits for u in self.U))
+
+    def pretty_name(self) -> str:
+        return f"B[{'⊗'.join(u.pretty_name()[2:-1] for u in self.U)}]"
+
+    @cached_property
+    def alpha(self) -> SymbolicFloat:
+        return reduce(lambda a, b: a * b.alpha, self.U, 1.0)
+
+    @cached_property
+    def num_ancillas(self) -> int:
+        return sum(u.num_ancillas for u in self.U)
+
+    @cached_property
+    def num_resource(self) -> int:
+        return sum(u.num_resource for u in self.U)
+
+    @cached_property
+    def epsilon(self) -> SymbolicFloat:
+        return sum(u.alpha * u.epsilon for u in self.U)
+
+    @property
+    def target_registers(self) -> Tuple[Register, ...]:
+        return (self.signature.get_right("system"),)
+
+    @property
+    def junk_registers(self) -> Tuple[Register, ...]:
+        return (self.signature.get_right("resource"),)
+
+    @property
+    def selection_registers(self) -> Tuple[Register, ...]:
+        return (self.signature.get_right("ancilla"),)
+
+    @property
+    def signal_state(self) -> PrepareOracle:
+        raise NotImplementedError
+
+    def build_composite_bloq(
+        self, bb: BloqBuilder, system: SoquetT, ancilla: SoquetT, resource: SoquetT
+    ) -> Dict[str, SoquetT]:
+        transpose = lambda x: zip(*x)
+        sys_regs, anc_regs, res_regs = transpose(
+            [evolve(r, name=f"{r.name}{i}") for r in u.signature.lefts()]
```

Comment:
I think this naming convention may require a suffix underscore (`_`). -- cc @fdmalone ran into an issue where the cirq interop adds a suffix in front of qubits due to which two different qubits can end up with the same name. This is the reason why the target registers in `QROM`s have a underscore in the suffix.

Reply:
is this still an issue? ideally we wouldn't be mutilating our signatures to support compiling to cirq

Reply:
I don't know if it is still an issue. We should verify and fix it before potentially creating new tech debt since we know we have encountered the issue in the past.

Reply:
I added the underscore suffix.